x-pack/plugins/security/server/lib/authentication/authentication_result.ts

```diff
@@ -73,13 +75,21 @@ export class AuthenticationResult {
   /**
    * Produces `AuthenticationResult` for the case when authentication fails.
    * @param error Error that occurred during authentication attempt.
+   * @param [challenges] Optional list of the challenges that will be returned to the user within
+   * `WWW-Authenticate` HTTP header. Multiple challenges will result into multiple headers (one per
+   * challenge) as it's better supported by the browsers than comma separated list within a single
+   * header. Challenges can only be set for errors with `401` error status.
    */
-  public static failed(error: Error) {
+  public static failed(error: Error, challenges?: string[]) {
```

Comment:
__note:__ If I understand that correctly `challenges` makes sense only with `401` error so I made it an optional argument for `failed`, but based on the way you handled it in your WIP I'd be glad to hear your thoughts or concerns!

Reply:
If I understand the current implementation correctly, when the Kerberos provider returns with `AuthenticationResult.failed(Boom.unauthorized(), ['Negotiate'])`, this will immediately cause the response to be sent to the user. Assuming the user's browser doesn't support Kerberos, will they still eventually get to the login screen?

The previous WIP approach which I implemented was intending to allow the `kerberos` provider to behave more seamlessly with the `basic` provider, which is why the challenges weren't necessarily terminating the way we iterate through the auth providers within the authenticator.

Other internals might have changed where the previous approach is no longer relevant.

Reply:
> If I understand the current implementation correctly, when the Kerberos provider returns with AuthenticationResult.failed(Boom.unauthorized(), ['Negotiate']), this will immediately cause the response to be sent to the user. Assuming the user's browser doesn't support Kerberos, will they still eventually get to the login screen?

Right, the only way user will be able to login using `basic` auth provider in `[kerberos, basic]` setup would be to go directly to `/login` page. Like we do for `saml`.

> The previous WIP approach which I implemented was intending to allow the kerberos provider to behave more seamlessly with the basic provider, which is why the challenges weren't necessarily terminating the way we iterate through the auth providers within the authenticator.

Thanks for pointing out! I've played with this approach last week, it looks interesting, but I'd rather tackle this separately if find out that current approach is annoying for our users. Would that work for you? There are three things that make me think about handling this separately:

* Same experience (okay, ^experience^ :slightly_smiling_face: ) even if `basic` is not enabled
* `basic` in `[kerberos, basic]` scenario feels more like admin-only and "normal" users that can't exchange Kerberos tickets to access tokens wouldn't benefit from `login` page too much, and admins know where to go
* `Authenticator` will have to handle `challenges` in a special manner, that I'd rather avoid unless it's really needed.

Reply:
I think what you have is perfectly fine :) I think at some point, we'll likely need to add the "choose your login method" when we have multiple auth providers configured.